pubsub/google/cloud/pubsub_v1/publisher/client.py

```diff
@@ -274,3 +274,22 @@ def publish(self, topic, data, **attrs):
                 batch = self._batch(topic, create=True)
 
         return future
+
+    def stop(self):
+        """Immediately publish all outstanding batches.
+
+        This asynchronously pushes all outstanding messages
+        and waits until all futures resolved. Method should be
+        invoked prior to deleting this Client object in order
+        to ensure that no pending messages are lost.
+
+        .. note::
+
+            This method blocks until all futures of all
+            batches resolved.
+        """
+        for topic in self._batches:
+            self._batches[topic].commit()
+
+        for topic in self._batches:
+            self._batches[topic].wait()
```

Comment:
@anguillanneuf, I've changed my plan after seeing Java's analogue. Now on `stop()` call all batches will start commiting in parallels (`commit()` creates a thread). After starting all of them, we'll just wait them in cycle. Still, we can return to sequential variant, if this is not an option

Reply:
`google.cloud.pubsub_v1.publisher._batch.thread.Batch` doesn't have a `wait` method.  Should we change its `commit` method to return the thread, and then wait on those threads?  If so, we also need to pass through a `timeout` there.

Reply:
@tseaver, I've took one more deeper look (as Tianzi said, Java is canonical for PubSub - let's build on this). I don't see timeouts in other languages. Thus, we shouldn't use them here, for consistency

I'm adding `wait()` into `Batch` with this PR actually. I'd say returning threads makes some sense, but every concrete thread related to a concrete `Batch`, so, I believe, it's better to keep threads in `Batch` objects. Seems to me, it's easier to understand `Batch.wait()`, than `join()` on a list of stray thread objects

(looks like I've deleted my comment and yours too, sorry for that)

Reply:
I also agree that returning threads is not as clean as keeping them within the Batch class. However, I wonder if we can return a future instead that the Client can keep track of. I am unsure when we'd check the status of this future when we're not waiting (a timer?) I'd love to hear your thoughts.

Reply:
Client.publish() already returns futures, so we expect clients to be using futures and waiting on those. But in the new Client.stop() method, we don't return a future and instead wait inside the method. We'd give the users more flexibility if we returned a future-like interface.

One option is to return the self._futures from Batch.commit(). Then, in Client.stop(), we take all of the returned futures from all the batches and then return them directly to the user to wait on. Pro: This option lets users choose when to wait and how long to do so. Cons: more complexity, exposing internal futures. 

Note that there's already a timeout for sending messages during publish. So if we don't let the user set a timeout for stop(), that timeout would still bound how long stop() takes. I don't know how many users would want to lengthen or shorten then stop() timeout beyond the publish timeout.

Reply:
If stop returns a future, it seems like the best thing would be for it to return a single future that is only fulfilled when all of the underlying futures are fulfilled. I think returning a list of futures would be counterintuitive. The user already would be holding on to the futures for the individual publishes themselves, so there probably isn't a need to return a bunch of lower-level futures.

Reply:
I'm fine with consistency with the other languages. Java doesn't return anything from stop. Given that on a publish, one can wait for all publishes to succeed or fail by waiting on the futures that came back from the publish calls, which is what the documentation for the Java shutdown method suggests. Therefore, returning a future on stop may be overkill. Is there something else we need stop() to do that requires it to wait on other threads or futures?

Reply:
The extra thing stop() does is to immediately trigger Batch.commit(). If one were to rely on waiting for the futures returned by Client.publish(), they will have to wait until the batch send timeout to trigger Batch.commit().

Do users care about this delay? I'm unsure. I suppose Client.stop() would be convenience more than enabling a new use-case.

Reply:
But they could call stop() and then wait on the futures returned by publish(), right? So they don't end up waiting for the timeout. I think that what stop is supposed to accomplish is to immediately send all outstanding messages and to disallow future calls to publish. Therefore, the user can know the state of all messages and shut down without having to worry about other publishes coming in in the meantime.

Reply:
@kamalaboulhosn, I'd sent [comment](https://github.com/googleapis/google-cloud-python/pull/9365#discussion_r338188020), but I didn't saw [yours one](https://github.com/googleapis/google-cloud-python/pull/9365#discussion_r338179397) yet. Thus, mine is little bit overdue.

> But they could call stop() and then wait on the futures returned by publish(), right?

Makes sense. So the final proposition is to make `stop()` non-blocking, right? And let users wait for resolving the futures from `publish()`, if they want to. I'm okay with it

Reply:
I'm fine with changing stop() to be non-blocking. We should be extra careful about documenting the method. It's easy to think "stop" means all requests die or that it's blocking.

Reply:
Okay, pushed new commit, in which tried to implement everything, we'd been talking about